### PR TITLE
External Infinispan as cache - Part 3

### LINKED ISSUE #28754: External Infinispan as cache - Part 3 / login failures cache
### Description

See Part 1: https://github.com/keycloak/keycloak/issues/28140

### Discussion

_No response_

### Motivation

See Part 1: https://github.com/keycloak/keycloak/issues/28140

### Details

Part 3 of the series `External Infinispan as cache`

This part will implement `UserLoginFailureProvider` which provides a remote only `loginFailures` cache.

### PR DESCRIPTION
**Pull request by pruivo**
Implementation of UserLoginFailureProvider using remote caches only.

Closes #28754

### CODE REVIEW

**Comment by ahus1** [comment]
We found that in main there was a typo with the profiles in the CI, and I pushed one change to fix it for your tests. Before, the profile wasn't actually active :scream:. 

Let's wait for the CI to complete.

File: server-spi-private/src/main/java/org/keycloak/models/UserLoginFailureProviderFactory.java [hunk 1]
```diff
@@ -16,11 +16,12 @@
...
 package org.keycloak.models;
 
+import org.keycloak.provider.EnvironmentDependentProviderFactory;
 import org.keycloak.provider.ProviderFactory;
 
 /**
  * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
  */
-public interface UserLoginFailureProviderFactory<T extends UserLoginFailureProvider> extends ProviderFactory<T>  {
+public interface UserLoginFailureProviderFactory<T extends UserLoginFailureProvider> extends ProviderFactory<T>, EnvironmentDependentProviderFactory {
```
Comment (ahus1) [inline server-spi-private/src/main/java/org/keycloak/models/UserLoginFailureProviderFactory.java:25]: I didn't check the previous commits, but now I notice: While this technically works, the interface `EnvironmentDependentProviderFactory` should go on each implementation, as it is not a feature of the `UserLoginFailureProviderFactory`, but of each implementation.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for the PR, it looks good overall. 

Some general comments: 

* When adding a copyright header to a new file, use the current date. There is actually an IntelliJ setting to do this automatically, and I'll share that in another PR.
* When adding new classes, take a moment to add a class level JavaDoc to describe what their intent is. I see it present on some classes, still missing for example in `BaseUpdater` and `RemoteChangeLogTransaction`

**Comment by pruivo** [comment]
@ahus1 What is the policy regarding the copyright? Is Keycloak requiring it in each file?

**Comment by ahus1** [comment]
@pruivo - yes, on each file. We are lacking an enforcement though.

**Comment by pruivo** [comment]
Disabled the following tests because we are unable to control time in the docker container:
* `org.keycloak.testsuite.model.AuthenticationSessionProviderTest#testExpiredAuthSessions`
* `org.keycloak.testsuite.oauth.AccessTokenTest#accessTokenCodeExpired`

Modified the following test due to the async nature of hot rod client events
* `org.keycloak.testsuite.model.UserSessionProviderTest#testRemoveUserSessionsByRealm`

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for the PR! See below for two comment. Thanks!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderTest.java [hunk 2]
```diff
@@ -322,14 +323,20 @@ public void testRemoveUserSessionsByRealm(KeycloakSession session) {
...
 
-        assertEquals(0, session.sessions().getUserSessionsStream(realm, session.users().getUserByUsername(realm, "user1"))
-                .count());
-        assertEquals(0, session.sessions().getUserSessionsStream(realm, session.users().getUserByUsername(realm, "user2"))
-                .count());
+        var user1 = session.users().getUserByUsername(realm, "user1");
+        var user2 = session.users().getUserByUsername(realm, "user2");
+
+        // TODO! [pruivo] to be removed when the session cache is remote only
+        // TODO! the Hot Rod events are async
```
[the remark this answers is not in the document]
Comment (pruivo) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderTest.java:332]: The `eventuallyEquals` can be removed after the `UserSessionProvider` is implemented to use the remote cache. Those TODOs are here to remind me to revert the code.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/AuthenticationSessionProviderTest.java [hunk 3]
```diff
@@ -212,10 +214,12 @@ public void testAuthenticationSessionRestart(KeycloakSession session) {
     @Test
     @ModelTest
     public void testExpiredAuthSessions(KeycloakSession session) {
+        // for some reason, Profile is always null.
```
[the remark this answers is not in the document]
Comment (pruivo) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/AuthenticationSessionProviderTest.java]: I'm pretty sure `Profile.isFeatureEnabled` is throwing an NPE too. Anyway, I've applied your suggestion.

**Review by ahus1 — APPROVED** [review]
Thank you - all my comments have been addressed. Once this is merged, it might be time time rebase this PR, possibly after we change to Protostream in main.

**Comment by pruivo** [comment]
thanks @ahus1 !

**Pull request merged**